### ISSUE TITLE
Async-Await issues with WPF

### ISSUE THREAD

**Issue by TomQv** [picoe/Eto#645]
I didn't investigate this yet into deep, its just a first impression.
I'm using lots of async-await-tasks, to create bitmaps, to fetch data from sqlite, to update data etc.
With Wpf I'm facing exceptions like "this object is blocked by another thread" or "cant update ui-element from non-ui thread". 
Same code works fine with WinForms and OSX so I'm a bit uncertain, if I did something wrong or if the Eto.Wpf implementation has problems with tasks.

**Comment by cwensley** [comment]
Hey @TomQv,

Thanks for the report.  I don't think this is a problem with async/await, but rather creating/using bitmaps from different threads which WPF has issues without going through extra steps to freeze the bitmap object first.

I assume you're using Eto 2.3?  I believe this was fixed in develop branch, commit 0b98aebadfa38c4bb785c4c1de53a933b4952449.

**Comment by TomQv** [comment]
I'm still having these problems, even with the latest commit.
All operation with Graphics are ending up in InvalidOperationException . (only WPF)

System.InvalidOperationException occurred
  HResult=-2146233079
  Message=Der aufrufende Thread kann nicht auf dieses Objekt zugreifen, da sich das Objekt im Besitz eines anderen Threads befindet.
  Source=WindowsBase
  StackTrace:
       bei System.Windows.Threading.Dispatcher.VerifyAccess()
       bei MS.Internal.TextFormatting.SimpleRun.Create(FormatSettings settings, CharacterBufferRange charString, TextRun textRun, Int32 cp, Int32 cpFirst, Int32 runLength, Int32 widthLeft, Int32 idealRunOffsetUnRounded, Double pixelsPerDip)
       bei MS.Internal.TextFormatting.SimpleTextLine.Create(FormatSettings settings, Int32 cpFirst, Int32 paragraphWidth, Double pixelsPerDip)
       bei MS.Internal.TextFormatting.TextFormatterImp.FormatLineInternal(TextSource textSource, Int32 firstCharIndex, Int32 lineLength, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak, TextRunCache textRunCache)
       bei MS.Internal.TextFormatting.TextFormatterImp.FormatLine(TextSource textSource, Int32 firstCharIndex, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak)
       bei System.Windows.Media.FormattedText.LineEnumerator.FormatLine(TextSource textSource, Int32 textSourcePosition, Double maxLineLength, TextParagraphProperties paraProps, TextLineBreak lineBreak)
       bei System.Windows.Media.FormattedText.LineEnumerator.MoveNext()
       bei System.Windows.Media.FormattedText.DrawAndCalculateMetrics(DrawingContext dc, Point drawingOffset, Boolean getBlackBoxMetrics)
       bei System.Windows.Media.FormattedText.Draw(DrawingContext dc, Point origin)
       bei System.Windows.Media.DrawingContext.DrawText(FormattedText formattedText, Point origin)
       bei Eto.Wpf.Drawing.GraphicsHandler.DrawText(Font font, SolidBrush b, Single x, Single y, String text) in C:\Users\Tom\QuoVadis\Qvx1\Eto\Eto-develop\Source\Eto.Wpf\Qvx\GraphicsHandler_qvx.cs:Zeile 314.
  InnerException:

**Comment by cwensley** [comment]
Hey, thanks for the stack trace, though unfortunately it doesn't give much information to go on.  Perhaps a little sample of what you're doing would help?

**Comment by cwensley** [comment]
Hm, guessing that perhaps it is either the font and/or brush objects if you are creating those from a different thread..

**Comment by TomQv** [comment]
yes:

System.InvalidOperationException occurred
  HResult=-2146233079
  Message=Der aufrufende Thread kann nicht auf dieses Objekt zugreifen, da sich das Objekt im Besitz eines anderen Threads befindet.
  Source=WindowsBase
  StackTrace:
       bei System.Windows.Threading.Dispatcher.VerifyAccess()
       bei MS.Internal.TextFormatting.SimpleRun.Create(FormatSettings settings, CharacterBufferRange charString, TextRun textRun, Int32 cp, Int32 cpFirst, Int32 runLength, Int32 widthLeft, Int32 idealRunOffsetUnRounded, Double pixelsPerDip)
       bei MS.Internal.TextFormatting.SimpleTextLine.Create(FormatSettings settings, Int32 cpFirst, Int32 paragraphWidth, Double pixelsPerDip)
       bei MS.Internal.TextFormatting.TextFormatterImp.FormatLineInternal(TextSource textSource, Int32 firstCharIndex, Int32 lineLength, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak, TextRunCache textRunCache)
       bei MS.Internal.TextFormatting.TextFormatterImp.FormatLine(TextSource textSource, Int32 firstCharIndex, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak)
       bei System.Windows.Media.FormattedText.LineEnumerator.FormatLine(TextSource textSource, Int32 textSourcePosition, Double maxLineLength, TextParagraphProperties paraProps, TextLineBreak lineBreak)
       bei System.Windows.Media.FormattedText.LineEnumerator.MoveNext()
       bei System.Windows.Media.FormattedText.DrawAndCalculateMetrics(DrawingContext dc, Point drawingOffset, Boolean getBlackBoxMetrics)
       bei System.Windows.Media.FormattedText.Draw(DrawingContext dc, Point origin)
       bei System.Windows.Media.DrawingContext.DrawText(FormattedText formattedText, Point origin)
       bei Eto.Wpf.Drawing.GraphicsHandler.DrawText(Font font, SolidBrush b, Single x, Single y, String text) in C:\Users\Tom\QuoVadis\Qvx1\Eto\Eto-develop\Source\Eto.Wpf\Drawing\GraphicsHandler.cs:Zeile 314.
  InnerException:

**Comment by TomQv** [comment]
There is still issues with this.
I realized this now with graphics-oerations, like 
graphics = new Eto.Drawing.Graphics(bmp); and then draw something.
Nothing is drawn, when all this happens in a async-await block.
Same code with Winforms is fine. 
I didn't work with eto for a while, just updated to the latest develop-commit, but as far as I remember, this was working 2 month ago. 
I don't get any exceptions btw.

**Comment by cwensley** [comment]
@TomQv thanks for the information!  Perhaps you're doing something like creating the graphics object in one thread then using it in another?  That would definitely not be a supported way of doing it.  

If you're not doing that, please file a new issue with a repro and I can look into it.

**Comment by TomQv** [comment]
I realized , that underlying GDI+ is not the right choice when it comes to multithreading rendering of graphics. Using OpenCV now for graphics and this fits much better to our use-case.
So I will not following this up any more..

**Comment by cwensley** [comment]
Hm, I'm not sure I follow since this is about WPF, but if it isn't an issue for you anymore let me know and I can close the issue.

**Comment by TomQv** [comment]
For some reason it came up with WPF, maybe because WPF is more strict in this. We had to change the design of our app for proper multi-threading bitmap-processing. 
You can close this. Thanks you.

**Comment by cwensley** [comment]
Thanks for following up. WPF is definitely much more strict, but I think if you use the graphics object in the same thread it should work.  I'm glad you were able to get it working for your app.
